ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,498 @@
+			SELECT note_resources.note_id as id
+			FROM note_resources
+			WHERE 0 UNION ${new Array(requiredResources.length).fill(noteIDsWithResource).join(' UNION ')}
+		)`;
+
+		const where = `
+		OR ROWID IN (
+			SELECT notes_normalized.ROWID
+			FROM notes_with_any_required_resources
+			JOIN notes_normalized
+			ON notes_with_any_required_resources.id=notes_normalized.id
+		)`;
+
+		withs.push(withRequired);
+		params.push(...requiredResources);
+		conditions.push(where);
+	}
+
+	if (relation === 'OR' && (excludedResources.length > 0)) {
+		const allNotesWithResources = `
+		all_notes_with_resources
+		AS (
+			SELECT DISTINCT note_resources.note_id AS id FROM note_resources
+		)`;
+		withs.push(allNotesWithResources); // for reuse
+
+		const notesWithoutExcludedResource = `
+		SELECT * FROM (
+			SELECT *
+			FROM all_notes_with_resources
+			EXCEPT ${noteIDsWithResource}
+		)`;
+
+		const withNoExcluded = `
+		notes_without_atleast_one_excluded_resource
+		AS (
+			${new Array(excludedResources.length).fill(notesWithoutExcludedResource).join(' UNION ')}
+		)`;
+
+		const where = `
+		OR ROWID IN (
+			SELECT notes_normalized.ROWID
+			FROM notes_without_atleast_one_excluded_resource
+			JOIN notes_normalized
+			ON notes_without_atleast_one_excluded_resource.id=notes_normalized.id
+		)`;
+		withs.push(withNoExcluded);
+		params.push(...excludedResources);
+		conditions.push(where);
+	}
+};
+
+const typeFilter = (filters: Term[], conditions: string[], relation: string) => {
+	const typeOfNote = filters.filter(x => x.name === 'type' && !x.negated).map(x => x.value);
+	typeOfNote.forEach(type => {
+		conditions.push(`
+		${relation} ROWID IN (
+			SELECT ROWID
+			FROM notes_normalized
+			WHERE notes_normalized.is_todo=${type === 'todo' ? 1 : 0}
+		)`);
+	});
+};
+
+const completedFilter = (filters: Term[], conditions: string[], relation: string) => {
+	const values = filters.filter(x => x.name === 'iscompleted' && !x.negated).map(x => x.value);
+	values.forEach(value => {
+		conditions.push(`
+		${relation} ROWID IN (
+			SELECT ROWID
+			FROM notes_normalized
+			WHERE notes_normalized.is_todo = 1
+			AND notes_normalized.todo_completed ${value === '1' ? '!= ' : '= '} 0
+		)`);
+	});
+};
+
+const getUnixMs = (date:string): string => {
+	const yyyymmdd = /^[0-9]{8}$/;
+	const yyyymm = /^[0-9]{6}$/;
+	const yyyy = /^[0-9]{4}$/;
+	const smartValue = /^(day|week|month|year)-([0-9]+)$/i;
+
+
+	if (yyyymmdd.test(date)) {
+		return moment.utc(date, 'YYYYMMDD').format('x');
+	} else if (yyyymm.test(date)) {
+		return moment.utc(date, 'YYYYMM').format('x');
+	} else if (yyyy.test(date)) {
+		return moment.utc(date, 'YYYY').format('x');
+	} else if (smartValue.test(date)) {
+		const match = smartValue.exec(date);
+		const timeUnit = match[1]; // eg. day, week, month, year
+		const num = Number(match[2]); // eg. 1, 12, 101
+		return moment.utc().startOf(timeUnit).subtract(num, timeUnit).format('x');
+	} else {
+		throw new Error('Invalid date format!');
```

Comment:
Shouldn't I be using the local date/time here instead of UTC?

`created:day-1` should return notes created from yesterday onwards in local time, right?

Reply:
Yes indeed all user-facing date/time should be in local time.